scp/cc/azure/attestation/src/sev_guest.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Portions Copyright (c) Microsoft Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <fcntl.h>
+#include <stdint.h>
+#include <sys/ioctl.h>
+#include <sys/types.h>
+
+#include <algorithm>
+#include <string>
+
+namespace google::scp::azure::attestation::sev_guest {
+
+#define SNP_GUEST_REQ_IOC_TYPE 'S'
+#define SNP_GET_REPORT               \
+  _IOWR(SNP_GUEST_REQ_IOC_TYPE, 0x0, \
+        struct google::scp::azure::attestation::sev::Request)
+#define SNP_GET_DERIVED_KEY          \
+  _IOWR(SNP_GUEST_REQ_IOC_TYPE, 0x1, \
+        struct google::scp::azure::attestation::sev::Request)
+#define SNP_GET_EXT_REPORT           \
+  _IOWR(SNP_GUEST_REQ_IOC_TYPE, 0x2, \
+        struct google::scp::azure::attestation::sev::Request)
+
+/* linux kernel 6.* versions of the ioctls that talk to the PSP */
+
+// aka/replaced by this from include/uapi/linux/sev-guest.h
+//
+struct Request {
+  uint8_t msg_version;  // message version number (must be non-zero)
+  uint64_t req_data;    // Request and response structure address
+  uint64_t resp_data;
+  uint64_t fw_err;  // firmware error code on failure (see psp-sev.h)
+};
+
+SnpReport* getReport(const std::string report_data) {
+  SnpRequest request = {};
+  auto decodedBytes =
+      utils::decodeHexString(report_data, sizeof(request.report_data));
+  size_t numBytesToCopy =
+      std::min(decodedBytes.size(), sizeof(request.report_data));
+  std::copy(decodedBytes.begin(), decodedBytes.begin() + numBytesToCopy,
+            request.report_data);
+
+  SnpResponse response = {};
+
+  Request payload = {
+      .msg_version = 1,
+      .req_data = (uint64_t)&request,
+      .resp_data = (uint64_t)&response,
+  };
+
+  auto sev_guest_file = open("/dev/sev-guest", O_RDWR | O_CLOEXEC);
+
+  auto rc = ioctl(sev_guest_file, SNP_GET_REPORT, &payload);
+  if (rc < 0) {
+    throw std::runtime_error("Failed to issue ioctl SNP_GET_REPORT");
```

Comment:
Just question to know the intention... Do we want to catch these exceptions or is it used to as "panic"?

Reply:
I'm open to suggestions about this, I think it's a fatal error to not be able to get the attestation so I don't think this library at least should handle the error. Presumably that logic also applies to the B&A code which calls this library, the question is more whether there's some convention of how to terminate in the event of a fatal error

Reply:
> Presumably that logic also applies to the B&A code which calls this library

Yeah, to be clear, if some code should catch it, it should be the B&A code.

The convention seems to be `CHECK` macro (similar to `DCHECK`).  It terminate the program with error message and the stack trace. In that sense it's similar to throw without catching.

Reply:
https://abseil.io/docs/cpp/guides/logging

It's part of the abseil's logging lib, so maybe it would be easier to setup log query in a monitor system like grafana.

Reply:
After having a think, I've decided to contain all errors raised by the attestation library, at the top level function, I return `std::optional` instead and it's `nullopt` in the case that there's an exception

Then I `CHECK()` the output of each call to the library to enforce that it has a value

Reply:
Sounds good. When we test it in a realistic Azure environment, let's make sure:
- `throws` behaves as intended. If (for example) `-fno-exceptions` is used , `CHECK` won't be called when exceptions are thrown.
- We can get enough information though log (or maybe opentelemetry) when a problem happens in production. If we can narrow down the cause before doing actual debugging work, it should be much nicer.